Tests/StoreKitUnitTests/StoreMessagesHelperTests.swift

```diff
@@ -0,0 +1,165 @@
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: false)
+
+        let message1 = MockStoreMessage(reason: Message.Reason.billingIssue)
+        let message2 = MockStoreMessage(reason: Message.Reason.priceIncreaseConsent)
+
+        try await waitForDeferredMessages(messages: [message1, message2])
+
+        await self.helper.showStoreMessages(types: [StoreMessageType.billingIssue])
+
+        expect(message1.displayCalled) == true
+        expect(message2.displayCalled) == false
+    }
+
+    @available(iOS 16.4, *)
+    func testShowMessagesAfterDeferMessagesAndShowingMessagesAutomaticallyDoesNotShowsMessages() async throws {
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: true)
+
+        let message1 = MockStoreMessage(reason: Message.Reason.billingIssue)
+        let message2 = MockStoreMessage(reason: Message.Reason.priceIncreaseConsent)
+
+        try await waitForDeferredMessages(messages: [message1, message2])
+
+        await self.helper.showStoreMessages(types: Set(StoreMessageType.allCases))
+
+        expect(message1.displayCalled) == false
+        expect(message2.displayCalled) == false
+    }
+
+    @available(iOS 16.4, *)
+    func testShowMessagesAfterDeferMessagesAndNoMessagesDoesNothing() async throws {
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: true)
+
+        try await waitForDeferredMessages(messages: [])
+
+        await self.helper.showStoreMessages(types: Set(StoreMessageType.allCases))
+    }
+
+    @available(iOS 16.4, *)
+    private func createHelper(showStoreMessagesAutomatically: Bool) {
+        self.helper = StoreMessagesHelper(systemInfo: self.systemInfo,
+                                          showStoreMessagesAutomatically: showStoreMessagesAutomatically,
+                                          storeMessagesProvider: self.storeMessagesProvider)
+    }
+
+    @available(iOS 16.4, *)
+    private func waitForDeferredMessages(messages: [StoreMessage]) async throws {
+        try await self.helper.deferMessagesIfNeeded()
+
+        try await Task.sleep(nanoseconds: DispatchTimeInterval.milliseconds(50).nanoseconds)
+
+        for message in messages {
+            self.storeMessagesProvider.updatesPublisher.send(message)
+
+            try await Task.sleep(nanoseconds: DispatchTimeInterval.milliseconds(50).nanoseconds)
+        }
+    }
+}
+
+@available(iOS 16.0, *)
+private final class MockStoreMessage: @unchecked Sendable, StoreMessage {
+
+    let reason: Message.Reason
+
+    init(reason: Message.Reason) {
+        self.reason = reason
+    }
+
+    var displayCalled = false
+    var displayCallCount = 0
+    @MainActor func display(in scene: UIWindowScene) throws {
```

Comment:
nitpick, but I think our style guide says to have a line break in between these
```suggestion
    var displayCallCount = 0
    
    @MainActor func display(in scene: UIWindowScene) throws {
```